prov/gni/src/gnix_mr_notifier.c

```diff
@@ -0,0 +1,245 @@
+		ret = -ret_errno;
+		goto err_exit;
+	}
+
+	if (get_user_delta_args.user_delta == NULL) {
+		GNIX_WARN(FI_LOG_MR, "kdreg get_user_delta is NULL\n");
+		ret = -FI_ENODATA;
+		goto err_exit;
+	}
+
+	mrn->fd = kdreg_fd;
+	mrn->cntr = (kdreg_user_delta_t *) get_user_delta_args.user_delta;
+
+err_exit:
+	fastlock_release(&mrn->lock);
+
+	return ret;
+}
+
+int
+_gnix_notifier_close(struct gnix_mr_notifier *mrn)
+{
+	int ret = FI_SUCCESS;
+	int ret_errno;
+
+	ret = notifier_verify_stuff(mrn);
+
+	if (ret == 0) {
+		fastlock_acquire(&mrn->lock);
+
+		if (close(mrn->fd) != 0) {
+			ret_errno = errno;
+			GNIX_WARN(FI_LOG_MR, "error closing kdreg device: %s\n",
+				  strerror(ret_errno));
+			// Not all of these map to fi_errno values
+			ret = -ret_errno;
+			goto err_exit;
+		}
+
+		mrn->cntr = NULL;
+	err_exit:
+		fastlock_release(&mrn->lock);
+	}
+
+	return ret;
+}
+
+static inline int
+kdreg_write(struct gnix_mr_notifier *mrn, void *buf, size_t len) {
+	int ret;
+
+	ret = write(mrn->fd, buf, len);
+	if ((ret < 0) || (ret != len)) {
+		// Not all of these map to fi_errno values
+		ret = -errno;
+		GNIX_WARN(FI_LOG_MR, "kdreg_write failed: %s\n",
+			  strerror(errno));
+		return ret;
+	}
+
+	return FI_SUCCESS;
+}
+
+int
+_gnix_notifier_monitor(struct gnix_mr_notifier *mrn,
+		    void *addr, uint64_t len, uint64_t cookie)
+{
+	int ret;
+	struct registration_monitor rm;
+
+	ret = notifier_verify_stuff(mrn);
+
+	if (ret == 0) {
+		GNIX_INFO(FI_LOG_MR, "monitoring %p (len=%lu) cookie=%lu\n",
+			  addr, len, cookie);
+
+		memset(&rm, 0, sizeof(rm));
+		rm.type = REGISTRATION_MONITOR;
+		rm.u.mon.addr = (uint64_t) addr;
+		rm.u.mon.len = len;
+		rm.u.mon.user_cookie = cookie;
+
+		ret = kdreg_write(mrn, &rm, sizeof(rm));
+	}
+
+	return ret;
+}
+
+int
+_gnix_notifier_unmonitor(struct gnix_mr_notifier *mrn, uint64_t cookie)
+{
+	int ret;
+	struct registration_monitor rm;
+
+	ret = notifier_verify_stuff(mrn);
+	if (ret == 0) {
+		GNIX_INFO(FI_LOG_MR, "unmonitoring cookie=%lu\n", cookie);
+
+		memset(&rm, 0, sizeof(rm));
+
+		rm.type = REGISTRATION_UNMONITOR;
+		rm.u.unmon.user_cookie = cookie;
+
+		ret = kdreg_write(mrn, &rm, sizeof(rm));
+	}
+
+	return ret;
+}
+
+int
+_gnix_notifier_get_event(struct gnix_mr_notifier *mrn, void* buf, size_t len)
+{
+	int ret, ret_errno;
+
+	if ((mrn == NULL) || (buf == NULL) || (len <= 0)) {
+		GNIX_WARN(FI_LOG_MR,
+			  "Invalid argument to _gnix_notifier_get_event\n");
+		return -FI_EINVAL;
+	}
+
+	if (*(mrn->cntr) > 0) {
+		GNIX_INFO(FI_LOG_MR, "reading kdreg event\n");
+		ret = read(mrn->fd, buf, len);
+		if (ret >= 0) {
+			return ret;
+		} else {
+			ret_errno = errno;
+			if (ret_errno != EAGAIN) {
+				GNIX_WARN(FI_LOG_MR,
+					  "kdreg event read failed: %s\n",
+					  strerror(ret_errno));
+			}
+			// Not all of these map to fi_errno values
+			return -ret_errno;
+		}
+	} else {
+		GNIX_INFO(FI_LOG_MR, "nothing to read from kdreg :(\n");
+		return -FI_EAGAIN;
+	}
+}
+
+#endif /* HAVE_KDREG */
```

Comment:
I agree.  I plan to have it run without kdreg for a bit before flipping the switch.  Then I'll have a small subset of tests run both with and without.